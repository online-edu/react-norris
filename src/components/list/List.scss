@import '../../styles/colors';
.jokes-list {
  .jokes-list__item {
    min-height: 64px;
    border-bottom: $list-divider;
  }
  .jokes-list__item--divider {
    border-bottom: $list-divider;
    &:last-child {
      border-bottom: none;
    }
  }
  .jokes-list__item-empty-emoticon {
    font-size: 8em;
  }
  .jokes-list__item-empty {
    color: $link;
  }
  .fade-enter {
    opacity: 0.01;
  }
  .fade-enter-active {
    opacity: 1;
    transition: opacity 500ms ease-in;
  }
  .fade-exit {
    opacity: 1;
  }
  .fade-exit-active {
    opacity: 0.01;
    transition: opacity 500ms ease-in;
  }
}
