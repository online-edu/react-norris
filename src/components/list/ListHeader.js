import React from 'react';
import PropTypes from 'prop-types';
import Switch from '../switch';
import Spinner from '../spinner';

/**
 * List header component.
 */
const ListHeader = ({
  title,
  loader,
  btnCaption: text,
  onBtnClick: click,
  switchCaption: svitch,
  onSwitchToggle,
}) => (
  <li className="list-group-item d-flex justify-content-between align-items-center jokes-list__item">
    <h5 className="mt-0 mb-1">{title}</h5>
    {loader && <Spinner />}
    {text && (
      <button type="button" className="btn btn-primary" onClick={click}>
        {text}
      </button>
    )}
    {svitch && (
      <Switch id={svitch} label={svitch} switchToggle={onSwitchToggle} />
    )}
  </li>
);
/**
 * List item default props.
 */
ListHeader.defaultProps = {
  loader: false,
  btnCaption: undefined,
  switchCaption: undefined,
  onSwitchToggle: () => {},
  onBtnClick: () => {},
};
/**
 * List item props types.
 */
ListHeader.propTypes = {
  /** Header title */
  title: PropTypes.string.isRequired,
  /** Button caption */
  btnCaption: PropTypes.string,
  /** Toggle spinner */
  loader: PropTypes.bool,
  /** Gets called when the button is clicked. */
  onBtnClick: PropTypes.func,
  /** Switch caption */
  switchCaption: PropTypes.string,
  /** Gets called when the switch is toggled. */
  onSwitchToggle: PropTypes.func,
};
export default ListHeader;
