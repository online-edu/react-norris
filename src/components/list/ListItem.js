import React from 'react';
import PropTypes from 'prop-types';
import FavButton from '../fav-button';

/**
 * List item component.
 */
const ListItem = ({ favorite, item, onFavoriteClick }) => (
  <li className="list-group-item d-flex justify-content-between align-items-center">
    {item.joke}
    <FavButton
      favorite={favorite}
      onClick={toggle => onFavoriteClick(toggle)}
    />
  </li>
);
/**
 * List item props types.
 */
ListItem.propTypes = {
  /** Toggle favorite */
  favorite: PropTypes.bool.isRequired,
  /** Gets called when the fav button is clicked. */
  onFavoriteClick: PropTypes.func.isRequired,
  /** Movie details */
  item: PropTypes.shape({ joke: PropTypes.string.isRequired }).isRequired,
};

export default ListItem;
