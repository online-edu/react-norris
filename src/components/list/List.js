import React from 'react';
import PropTypes from 'prop-types';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import ListHeader from './ListHeader';
import ListItem from './ListItem';
import ListEmpty from './ListEmpty';
import './List.scss';

/**
 * List component.
 */
const List = ({
  btnCaption,
  title,
  items = [],
  loader,
  favorite,
  onHeaderAction,
  onFavoriteClick,
}) => (
  <ul className="list-group jokes-list">
    <TransitionGroup>
      <ListHeader
        title={title}
        loader={loader}
        btnCaption={btnCaption}
        onBtnClick={onHeaderAction}
      />
      {items
        && items.map(item => (
          <CSSTransition key={item.id} timeout={500} classNames="fade">
            <ListItem
              item={item}
              favorite={favorite}
              onFavoriteClick={fav => onFavoriteClick({ item, fav })}
            />
          </CSSTransition>
        ))}
      {items.length === 0 && (
        <CSSTransition timeout={100} classNames="fade">
          <ListEmpty />
        </CSSTransition>
      )}
    </TransitionGroup>
  </ul>
);
/**
 * List default props.
 */
List.defaultProps = {
  loader: false,
  favorite: false,
  btnCaption: undefined,
  onHeaderAction: () => {},
};
/**
 * List props types.
 */
List.propTypes = {
  /** Header title */
  title: PropTypes.string.isRequired,
  /** Button caption */
  btnCaption: PropTypes.string,
  /** Toggle spinner */
  loader: PropTypes.bool,
  /** Gets called when the header button is clicked. */
  onHeaderAction: PropTypes.func,
  /** Gets called when the fav button is clicked. */
  onFavoriteClick: PropTypes.func.isRequired,
  /** List of items */
  items: PropTypes.array.isRequired,
  /** Fav or not fav visibility */
  favorite: PropTypes.bool,
};

export default List;
