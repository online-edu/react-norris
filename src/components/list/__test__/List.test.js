import React from 'react';
import { shallow, mount } from 'enzyme';
import List from '../List';
import ListHeader from '../ListHeader';
import ListItem from '../ListItem';

describe('<List />', () => {
  const onHeaderAction = jest.fn();
  const onFavoriteClick = jest.fn();
  const joke = [
    {
      id: 480,
      joke: 'The class object inherits from Chuck Norris',
      categories: ['nerdy'],
    },
  ];
  const wrapper = (
    <List
      title="Jokes"
      btnCaption="View more"
      loader={false}
      items={joke}
      onHeaderAction={onHeaderAction}
      onFavoriteClick={onFavoriteClick}
    />
  );

  it('renders correctly', () => {
    const component = shallow(wrapper);
    expect(component).toMatchSnapshot();
  });

  it('renders a <ListHeader /> component', () => {
    const component = shallow(wrapper);
    expect(component.find(ListHeader)).toHaveLength(1);
  });

  it('renders a <ListItem /> component', () => {
    const component = shallow(wrapper);
    expect(component.find(ListItem)).toHaveLength(1);
  });

  it('allows us to set props', () => {
    const component = mount(wrapper);
    component.setProps({ items: [] });
    expect(component.props().items.length).toBeLessThan(1);
  });
});
