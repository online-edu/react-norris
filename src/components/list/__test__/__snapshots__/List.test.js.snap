// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> renders correctly 1`] = `ShallowWrapper {}`;
